drake/systems/plants/RigidBodyManipulator.h

```diff
@@ -97,46 +99,256 @@ class DLLEXPORT_RBM RigidBodyManipulator
   std::string getVelocityName(int velocity_num) const;
   std::string getStateName(int state_num) const;
 
-  template <typename DerivedQ, typename DerivedV>
-  void doKinematics(const MatrixBase<DerivedQ> &q, const MatrixBase<DerivedV> &v, bool compute_gradients = false, bool compute_JdotV = true);
+  template <typename DerivedQ>
+  KinematicsCache<typename DerivedQ::Scalar> doKinematics(const MatrixBase<DerivedQ>& q, int gradient_order) {
+    KinematicsCache<typename DerivedQ::Scalar> ret(bodies, gradient_order);
+    ret.initialize(q);
+    doKinematics(ret);
+    return ret;
+  }
 
-  bool isBodyPartOfRobot(const RigidBody& body, const std::set<int>& robotnum);
+  template <typename DerivedQ, typename DerivedV>
+  KinematicsCache<typename DerivedQ::Scalar> doKinematics(const MatrixBase<DerivedQ>& q, const MatrixBase<DerivedV>& v, int gradient_order = 0, bool compute_JdotV = true) {
+    KinematicsCache<typename DerivedQ::Scalar> ret(bodies, gradient_order);
+    ret.initialize(q, v);
+    doKinematics(ret, compute_JdotV);
+    return ret;
+  }
 
-  double getMass(const std::set<int>& robotnum = RigidBody::defaultRobotNumSet);
+  template <typename Scalar>
+  void doKinematics(KinematicsCache<Scalar>& cache, bool compute_JdotV = false) const {
```

Comment:
Yes, but that doesn't mean that inline expansion will happen. This is my current understanding:
* the inline keyword doesn't actually command the compiler to use 'inline expansion' (i.e. putting the code of the function where the call happens, getting rid of the function call. At best, it's a suggestion to the compiler (https://msdn.microsoft.com/en-us/library/bw1hbe6y.aspx, https://isocpp.org/wiki/faq/Inline-Functions#overview-inline-fns). Whether or not the compiler will perform inline expansion depends on e.g. the size of the method.
* effects of the inline keyword on performance are very case-dependent (https://isocpp.org/wiki/faq/inline-functions#inline-and-perf)
* defining a method in the class body itself (in the header) automatically tells the compiler to make the method inline (as in, it does the same thing as adding the inline keyword) (https://isocpp.org/wiki/faq/Inline-Functions#inline-member-fns-more). This does *not* mean however that the compiler is forced to inline expand the code. It just follows the same rules as for any other function that is declared inline (https://stackoverflow.com/questions/5431362/how-does-templates-work-are-they-always-inlined)